[SYCL] Fix race that occurs when submitting to single queue in parallel

alexanderfle opened the pull request:
This PR aims to fix the race when setting kernel parameters in parallel. That occurs when submitting to single queue by multiple threads.

- In cases when we use cacheable kernel we must use mutex, as we share the same instance by threads.
- In cases when we don't use cacheable kernel each thread has its own instance of the kernel and so race-condition doesn't occur.

Regression: queue submit. single queue

Signed-off-by: Alexander Flegontov <alexander.flegontov@intel.com>

s-kanaev commented on sycl/source/detail/scheduler/scheduler.cpp:
```diff
@@ -83,7 +83,8 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
 
   {
-    std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
+    std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+    lockSharedTimedMutex(Lock);
```
Why?
We only need a read-lock to enqueue smth as it's non-modifying operation.

alexanderfle replied:
We have already a test for the test case, which sporadically fails without that on Windows and Linux.
About a couple of weeks ago, we had a single lock guard in the function for NewCmd and enqueue, but now, test fails when you try to submit to a single queue by multi-threads. 
This patch fixes the issue.

s-kanaev replied:
OK. You're going to serialize enqueue operations, aren't you? Then, I believe there should be a distinct lock for it. Justification: there are other places where we acquire read-lock and enqueue smth or do smth.

As far as I understand, in general there are two operation for graph: put command to graph and enqueue. We can't put multiple commands to graph simultaneously as it's write operation. Though we should be able to enqueue multiple commands for execution, isn't it?

alexanderfle replied:
Yes, I am going to serialize enqueue operations, because without it we can face with fails when try to submit to a single queue by multiple threads. Thanks, I will add a separate lock for that.

alexanderfle replied:
It was decided to use a single lock. see conversation below.

romanovvlad commented on sycl/source/detail/scheduler/scheduler.cpp:
```diff
@@ -83,7 +83,8 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
+    std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+    lockSharedTimedMutex(Lock);
 
     // TODO: Check if lazy mode.
```
You need to check that NewCmd on line 91 is still valid as it can be executed and removed between locks.
Suggest storing an event instead of command on line 66, then extract and check that it is non-null the command from the event on line 90.

Also line 97 looks not thread safe.

alexanderfle replied:
Ok, I will do that.

alexanderfle replied:
As we decided to use a single lock in the function Scheduler::addCG(), we no longer need to check NewCmd on line 91.

I checked line 97,  flushStreams() calls Scheduler::addHostAccessor() inside itself that locks MGraphLock.

romanovvlad commented:
Also, could you please develop a test for this?

s-kanaev commented on sycl/source/detail/scheduler/scheduler.cpp:
```diff
@@ -80,22 +81,27 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
   {
-    std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
+    std::unique_lock<std::shared_timed_mutex> Lock(MGraphLock, std::defer_lock);
+    lockSharedTimedMutex(Lock);
```
On the other hand, you don't need to unlock and lock again afterwards if you're going to acquire the write lock again.

alexanderfle replied:
Did you mean that it would be better to have a single lock that will be locked once as it was before?

s-kanaev replied:
OK, I'll clarify what I mean.
There are two possible approaches:
- employ a existing read-lock + serialization lock;
- employ existing write-lock.

As far as I understood the problem is that 'enqueue command for execution' is potentially modifying operation.

The first approach has an overhead, that's why we would prefer the second one.
Here, at line 83 the write-lock is dropped. Afterwards, write-lock is acquired again at line 89. Though nothing happens in current thread between these lines. Thus I'd prefer to not dropping the write-lock as I don't see any need to 'pause' the 'write-lockness' of the graph.

Also, there are other places where enqueue takes place. We need to change locking there too.

alexanderfle replied:
I tried to use a single lock in this function and change locking in other places where enqueue takes place, and I see that it leads to failures in some Lit tests. 

The same problem as for commit above "939dc81" - http://ci.llvm.intel.com:8010/#/builders/2/builds/3045
```
terminate called after throwing an instance of 'std::system_error'
  what():  Resource deadlock avoided
```
It requires deep investigation,
So probably, two separate locks are a hotfix.

alexanderfle commented:
@s-kanaev , @romanovvlad  I addressed your comments and applied as much as possible to pass all the Lit tests without fails and hangs. Could you please take a look and approve the current version of the patch?

s-kanaev commented on sycl/source/detail/scheduler/scheduler.cpp:
```diff
@@ -80,10 +80,6 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
     default:
       NewCmd = MGraphBuilder.addCG(std::move(CommandGroup), std::move(Queue));
     }
-  }
-
-  {
-    std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
```
There are also some other places which enqueue commands for execution. Currently they are using read-lock. I believe we can face the same issue there as well. Also, your intent here is to serialize enqueue operations. Well, the target of your intent is not achieved yet as enqueue operations may be executed in parallel in that places.

alexanderfle replied:
Yes, I checked all other places which enqueue commands for execution and faced problem the Lit tests hangs, you can have a look at previous commits and build bot results.
So, the Current version passed all Lit tests and fixed the problem with the regression test.

s-kanaev replied:
I believe there must be some solution here.
As far as I understood the changes have caused a deadlock. What's the purpose of write-lock? Is it employed here to prevent simultaneous changes to structure of graph or changes to some internal data of commands in graph? We can get by the latter with employing a per-command mutex or atomic property with proper combination of CAS/read/exchange operations.

alexanderfle replied:
Yes, I investigated the root cause of the problem and found the race when setting kernel parameters in parallel. That occurs when submitting to a single queue by multiple threads. So I added per-kernel mutex.

alexanderfle commented:
We already have a test for this, at the moment, which regresses on Linux and Windows and this patch fixes this problem.

s-kanaev commented on sycl/source/detail/scheduler/scheduler.cpp:
```diff
@@ -84,7 +84,6 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
 
   {
     std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
-
```
Unneeded change

alexanderfle replied:
done

s-kanaev commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -1799,15 +1799,19 @@ cl_int ExecCGCommand::enqueueImp() {
+          detail::ProgramManager::getInstance().getOrCreateKernel(
+              ExecKernel->MOSModuleHandle, Context, ExecKernel->MKernelName,
+              nullptr);
+      KernelMutex->lock();
```
It's much better to employ a `unique_lock` or `lock_guard` as they will call `unlock()` automatically when the object (`unique_lock` or `lock_guard`) is destroyed, e.g. upon exception thrown.

alexanderfle replied:
I agree I will do it.

s-kanaev commented on sycl/source/detail/kernel_program_cache.hpp:
```diff
@@ -58,8 +58,15 @@ class KernelProgramCache {
   using PiKernelT = std::remove_pointer<RT::PiKernel>::type;
+
+  struct BuildResultKernel : public BuildResult<PiKernelT> {
+    std::mutex MKernelMutex;
```
As far as I remember @sergey-semenov did something similar, see intel/llvm#1204. I bet he introduced per-program mutex which is locked when program builds. @sergey-semenov , will your change be unneeded if this one is commited?

sergey-semenov replied:
I don't think so. My change dealt specifically with attempting to call `sycl::program` API in parallel, which doesn't seem related to this change.

alexanderfle replied:
I agree with @sergey-semenov.

s-kanaev commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -1797,71 +1797,87 @@ cl_int ExecCGCommand::enqueueImp() {
+      std::lock_guard<std::mutex> Lock(*KernelMutex);
+      Error = SetKernelParamsAndLaunch();
+    } else {
+      Error = SetKernelParamsAndLaunch();
```
As far as I understand, we still will get a race-condition if we submit the same kernel with `nullptr != ExecKernel->MSyclKernel` in multiple threads, isn't it?
Is this case covered by some tests?

alexanderfle replied:
Right, I covered this case: `nullptr != ExecKernel->MSyclKernel`.
No, at the moment, but I will add the test in our test suite nearby the previous regression test.
So in the scope of the PR, I have done all.

alexanderfle commented:
@s-kanaev , @kbobrovs ,  in the scope of the PR, I have done all. Could you please review it?

alexanderfle commented:
@kbobrovs , We already have one regression test fails in our tests suit without this fix, and also I added an additional regression test in separate PR in our test suit, where I covered one case mentioned above as `nullptr != ExecKernel->MSyclKernel`. 
So, I believe that here I have done all.

kbobrovs commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -1797,71 +1797,100 @@ cl_int ExecCGCommand::enqueueImp() {
+              nullptr);
+    }
+
+    auto SetKernelParamsAndLaunch = [this, &ExecKernel, &Kernel, &NDRDesc,
```
Nit:  lambdas are supposed to be short lightweight inline functions, otherwise, like in this case, it rather hinders understanding. This would ideally be a separate function. Plus instead of duplicating the call below
```
    if (KernelMutex != nullptr) {
        // For cacheable kernels, we use per-kernel mutex
        std::lock_guard<std::mutex> Lock(*KernelMutex);
      Error = SetKernelParamsAndLaunch();
    } else {
      Error = SetKernelParamsAndLaunch();
```
mutex could be a parameter, non-null meaning that the lock needs to be acquired

alexanderfle replied:
Thanks for the comment, this is a good idea: replace lambda with a separate function. Regarding passing a mutex as a parameter,  I believe it is not a good idea because then I don't see how to use `std::lock_guard` without duplication of whole code of the function and using `lock()`, `unlock()` - as discussed above is bad practice. (this is a loss of automatically lock/unlock when the object is constructed/destroyed)

kbobrovs commented:
> @kbobrovs , We already have one regression test fails in our tests suit without this fix, and also I added an additional regression test in separate PR in our test suit, where I covered one case mentioned above as `nullptr != ExecKernel->MSyclKernel`.
> So, I believe that here I have done all.

I could not find which test was failing in this PR. Would be good to mention then in the description.

alexanderfle commented:
@kbobrovs , Could you please take a look again and approve?

alexanderfle commented:
@kbobrovs , ping

alexanderfle commented:
@bader , Could you please merge?

bader commented:
@alexanderfle, please, fix CI failures.

alexanderfle commented:
@kbobrovs , ping

alexanderfle commented:
@bader , ping
